src/database/adatabase.cpp

```diff
@@ -496,7 +496,8 @@ const QStringList ADatabase::getCompletionList(ADatabaseTarget target)
     return completer_list;
 }
 
-const QMap<QString, int> ADatabase::getIdMap(ADatabaseTarget target)
+const
+QMap<ADatabase::QueryResult_qstr, ADatabase::QueryResult_i> ADatabase::getIdMap(ADatabaseTarget target)
```

Comment:
`const QMap<QString, int> ADatabase::getIdMap(ADatabaseTarget target)`

Reply:
Can you give me examples of functions that cast the qvariant to other types because i still believe a QueryResult_t will make the fact that a query occurs visible at a glance

Reply:
Pretty much all of them do, here are 3 examples:
ADatabase::getIdMap
```
    while (query.next()) {
        id_map.insert(query.value(1).toString(), query.value(0).toInt());
    }
```
ADatabase::getLastEntry
```
        return query.value(0).toInt();
```
ADatabase::getForeignKeyConstraints
```
    while (query.next()) {
        row_ids.append(query.value(0).toInt());
    }
    return row_ids;
```

[QSqlQuery::value](https://doc.qt.io/qt-5/qsqlquery.html#value) always returns a QVariant, which is then cast to the needed type (in most functions) or directly written to the QMap<QString, QVariant> aka QMap<ColName, ColData> for the entries.

Reply:
Yes but the thing is that each function casts it to its specific type so QVariant returns from the db but these functions cast it to something "meaningful"
For each of them what is that?

Reply:
getIdMap returns the maps we talked about earlier, for example `QMap<PilotName, RowId> pilotsIdMap;`
getLastEntry returns an `int` aka `RowId`
getForeignKeyConstraints returns a `List<int>` aka `List<RowId>`

Let's say you're interested in one of these maps, you look how it is created and get to the database function - wouldn't it be easier to see the types the map is made of, `<QString, int>` than seeing a custom database type, then following that type to see its qstring and int? Since the maps are already named with aliases, having an aliased middle step feels counter-intuitive.

I have looked through the database functions and only one of them, `disconnect` does not query the database. So in a manner of speaking, you are getting a 'QueryResult_t` when you are using the ADatabase class, which seems clear enough. 

Also, having too many custom names obscures the underlying types. I'm all for using aliases to give meaningful names to otherwise convoluted complex data types, but I am not so eager about giving names to basic data types. There are of course arguments to be made, and I think having an int called RowId is great, since it is used like that in many cases and makes it easier to understand what it is for. Nevertheless, I also like to see at first glance what the underlying type is, when dealing with basic tasks.

Reply:
If you have forgotten what "IdMap" means  then seeing just `QMap<QString, int>` doesnt help that much. The aliasing starts from the function singature so that it ends in keeping you sane. First you see that it returns eg `QMap<PilotName_T, RowId_T>` and you either remember on the fly what IdMap means or you continue to look at the function body. 

You said it yourself that checking the type is as easy as hovering the mouse. The documentation can have the basic types so that it gives as much info as possible.
```c++
/// \brief Query the db and return pilotname and corresponding rowid as QString and int.
QMap<PilotName_T, RowId_T> getMap(ADatabaseTarget t);
```

Reply:
are we talking about the same thing? I was referring to having a seperate `QueryResult_t` as being unnecessary, naming the maps themselves is fine.

`QMap<QString, int> getIdMap` should return the basic types because we name them later on, once the maps are created, here it would be infeasible, since the return from getIdMap, <QString, int> can be a variety of maps, for example:

```cpp
QMap<PilotName_T, PilotRowId_T> pilotsIdMap;
QMap<TailRegistration_T, TailId_T> tailsIdMap;
QMap<AirportICAO_T, AirportId_T> airportIcaoIdMap;
QMap<AirportIATA_T, AirportId_T> airportIataIdMap;
QMap<AirportName_T, AirportId_T> airportNameIdMap;
```

Adding another intermediate type only obfuscates the underlying type, since then, if you would 'hover your mouse' over `PilotName` (or look it up), it would show you `QueryResult_t` instead of `QString`,  which is the piece of information you're after.

In general, I believe we should only use aliases when there is a good reason to do so. If every `int` and `QString` end up being their own custom type, it makes keeping an overview more difficult, not less (at least for me)

Reply:
We are talking about the same thing. If you see `getIdMap` return `QString` and `int` and take an argument `AdatabaseTarget` how can you at a glance infer that if you pass the pilot target its `PilotName_T`, `PilotRowId_T` and not `PilotCallsign_T`?

There will have to be a couple of logic steps you have to take anyway. Semantically the getIdMap fucntion queries the db based on the target and casts the return variant to string. So it makes sense to be abstractly aliased as `QueryResult_T` (step 1). And after that you see what is aliased as `QueryResult_T`(step 2). All that if need (which will be in the future).
All that obviously is semantic aliasing so anyone taking the returned QMap has a string in their hand. If we wanted *absolute* control of the return type of course we would have to make a custom one but thats unecessary (because again we want to use a string *but* each string has meaning for the program)

I dont know why we have one one such function and we dont overload it which will pretty much bring an end to this discussion. I can do it in 30 minutes because its just breaking up the switch into overloads and calling the proper ones around the program.

Reply:
> how can you at a glance infer that if you pass the pilot target its PilotName_T, PilotRowId_T and not PilotCallsign_T?

From there, the only thing you can infer is that you are getting a string. When you look at the map, that's where you see what the data is. But knowing you are getting a string is better than knowing you are getting some `QueryResult_T` that could be anything.
> All that obviously is semantic aliasing so anyone taking the returned QMap has a string in their hand.

Exactly. That's why I prefer having `QString` over an alias, `QueryResult_T` tells me nothing and creating an alias here is obfuscating the fact that it's no black magic we're dealing with here, just a good old string.
> There will have to be a couple of logic steps you have to take anyway.

One more reason to have as little steps as possible. Series of aliases add levels of abstraction that make it impossible to see at a glance what kind of data we are dealing with here.